### PR TITLE
Consume ITlsTokenBindingFeature in CookieAuthMiddleware.

### PR DESCRIPTION
**Pull request by Tratcher**
#263 @HaoK @PinpointTownes

I also manually verified it works with IE (Win10) and WebListener.

### CODE REVIEW

File: src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs [hunk 1]
```diff
@@ -357,5 +358,20 @@ private static bool IsHostRelative(string path)
...
         }
+
+        private string GetTlsTokenBinding()
+        {
+            string userTokenBinding = null;
+            var tokenBinding = Context.Features.Get<ITlsTokenBindingFeature>();
+            if (tokenBinding != null)
+            {
+                var binding = tokenBinding.GetProvidedTokenBindingId();
+                if (binding != null)
```
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs]: nit: collapse outer layer of if/null checks?
```
var binding = Context.Features.Get<ITlsTokenBindingFeature>()?.GetProvidedTokenBindingId();
```

Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs]: Could then also follow it up with no ifs and removing the local 
```
return binding == null ? null : Convert.ToBase64String(binding);
```

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs]: well if you want to be all fancy .... :-)

File: src/Microsoft.AspNet.Authentication/DataHandler/SecureDataFormat.cs [hunk 2]
```diff
@@ -20,15 +18,30 @@ public SecureDataFormat(IDataSerializer<TData> serializer, IDataProtector protec
...
-            string protectedText = Base64UrlTextEncoder.Encode(protectedData);
-            return protectedText;
+
+            var protector = _protector;
+            if (!string.IsNullOrEmpty(purpose))
+            {
+                protector = protector.CreateProtector(purpose);
+            }
+
+            byte[] protectedData = protector.Protect(userData);
```
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/DataHandler/SecureDataFormat.cs]: nit: var police are coming...

File: src/Microsoft.AspNet.Authentication/DataHandler/SecureDataFormat.cs [hunk 3]
```diff
@@ -42,7 +55,13 @@ public TData Unprotect(string protectedText)
...
                 }
 
-                byte[] userData = _protector.Unprotect(protectedData);
+                var protector = _protector;
+                if (!string.IsNullOrEmpty(purpose))
+                {
+                    protector = protector.CreateProtector(purpose);
+                }
+
+                byte[] userData = protector.Unprotect(protectedData);
```
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/DataHandler/SecureDataFormat.cs]: nit^3: var^2

**Comment by HaoK** [comment]
If you say it works, I guess I have to believe you :)  :shipit:

**Comment by PinpointTownes** [comment]
It looks really nice :+1:

**Comment by PinpointTownes** [comment]
Curious: are there any downsides using this TLS binding token stuff?

**Comment by Tratcher** [comment]
There may be some perf overhead, but I was promised it would be minimal. We'll find out...

**Pull request merged**